contracts/campaigns/KOLReferral.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.0;
+
+import "@opengsn/contracts/src/ERC2771Recipient.sol";
+
+contract KOLReferral is ERC2771Recipient {
+    mapping(address => string) public walletToReferralId;
+    mapping(string => address) public referralIdToWallet;
+    mapping(address => string) public userReferredBy;
+    address[] private _users;
+    address public admin;
+
+    modifier onlyAdmin() {
+        require(_msgSender() == admin, "Only Admin");
+        _;
+    }
+
+    constructor(address admin_, address trustedForwarder_) {
+        require(admin_ != address(0), "zero address");
+        require(trustedForwarder_ != address(0), "zero address");
+        admin = admin_;
+        _setTrustedForwarder(trustedForwarder_);
+    }
+
+    function registerKOL(address wallet, string memory referralId)
+        external
+        onlyAdmin
+    {
+        require(
+            referralIdToWallet[referralId] == address(0),
+            "ReferralId is already taken"
+        );
+        require(
+            bytes(walletToReferralId[wallet]).length == 0,
+            "ReferralId is already assigned for this wallet"
+        );
+        walletToReferralId[wallet] = referralId;
+        referralIdToWallet[referralId] = wallet;
+    }
+
+    // TODO: use msgSender() for gasless transaction
+    function storeUserInfo(string memory referralId) external {
+        require(
+            referralIdToWallet[referralId] != address(0),
+            "Invalid ReferralId"
+        );
+        require(
+            bytes(userReferredBy[_msgSender()]).length == 0,
+            "User is already referred before"
+        );
+        userReferredBy[_msgSender()] = referralId;
+        _users.push(_msgSender());
+    }
+
+    function queryUserReferrer(address user)
+        external
+        view
+        returns (address _referrer)
+    {
+        require(bytes(userReferredBy[user]).length != 0, "User not referred");
```

Comment:
sure will update.
Why returning address(0) is a better way than reverting?
Shall we check with @abhi-sdlabs on how he wants it?

Reply:
User not referred is not an error, its a genuine condition. 
public function userReferredBy(userAddress) gives you empty string, not an error. 
similarly this function, which is an extension of userReferredBy, should also, give default value rather than error.